File: compression/techniques/visual-basic/rle.vb
' Raspberry Pi Foundation
' Usage licensed under CC BY-SA 4
' Note: This file is designed to be copied out and compiled on your machine.
' To run this file you need to:
' 1. Copy the contents
' 2. Paste them into the Visual Basic IDE of your choice (Visual Studio, for example)
' 3. Compile the program
' 4. Run the program


Module Program

    ' The Main subroutine is the default entry point for a VB program
    Sub Main()
        Dim text As String = "aaaabbbccdddeee" ' First piece of test data
        Console.WriteLine(text)
        Console.WriteLine(RLE(text))

        text = "Bobby"
        Console.WriteLine(text)
        Console.WriteLine(RLE(text))

        text = "a"
        Console.WriteLine(text)
        Console.WriteLine(RLE(text))

        text = "aaaaaa"
        Console.WriteLine(text)
        Console.WriteLine(RLE(text))

        text = "aaaae"
        Console.WriteLine(text)
        Console.WriteLine(RLE(text))

        text = "aeeeee"
        Console.WriteLine(text)
        Console.WriteLine(RLE(text))

        Console.WriteLine("Finished")
        Console.ReadLine()
    End Sub

    Function RLE(ByVal textString As String) As String
        Dim currentToken As Char = textString(0)
        Dim compressed As String = currentToken.ToString()
        Dim counter As Integer = 1

        For i As Integer = 1 To textString.Length - 1
            Dim nextToken As Char = textString(i)

            If nextToken <> currentToken Then
                compressed = compressed & counter.ToString() & nextToken
                counter = 1
            Else
                counter = counter + 1
            End If

            currentToken = nextToken
        Next

        compressed = compressed & counter.ToString()
        Return compressed
    End Function

End Module

File: compression/techniques/java/rle.java
/*
Raspberry Pi Foundation
Developed as part of Ada Computer Science

Usage licensed under CC BY-NC-SA 4.0

Note: This file is designed to be copied out and compiled on your machine.
In order for the program to compile properly in some IDEs, you need to ensure that the
class filename is the same as the class name with the main method in it.

To run this file you need to:
1. Copy the contents
2. Create a Java project in the IDE of your choice (Eclipse, for example)
3. Create a new Java Class that uses the same name as the class containing the main method in this program
4. Paste the contents into the new Java Class
5. Save and run the program
*/

class CompressionTechniques
{
    // The main method is the entry point for all Java programs
    public static void main(String[] args) {
        // Test data - normal data
        String text = "aaaabbbccdddeee";
        System.out.println(text);
        System.out.println(rle(text));

        text = "Bobby";
        System.out.println(text);
        System.out.println(rle(text));

        // Test data - boundary data
        text = "a";
        System.out.println(text);
        System.out.println(rle(text));

        text = "aaaaaa";
        System.out.println(text);
        System.out.println(rle(text));

        text = "aaaae";
        System.out.println(text);
        System.out.println(rle(text));

        text = "aeeeee";
        System.out.println(text);
        System.out.println(rle(text));
    }


    // A simple version of Run Length Encoding for text strings
    public static String rle(String textString) {
        char currentToken = textString.charAt(0);
        String compressed = Character.toString(currentToken);
        int counter = 1;

        for (int i = 1; i < textString.length(); i++) {
            char nextToken = textString.charAt(i);
            if (nextToken != currentToken) {
                compressed = compressed + Integer.toString(counter) + nextToken;
                counter = 1;
            }
            else {
                counter = counter + 1;
            }
            currentToken = nextToken;
        }
        compressed = compressed + Integer.toString(counter);
        return compressed;
    }


}